mods/default/nodes.lua

```diff
@@ -586,29 +586,124 @@ minetest.register_node("default:snow", {
+	walkable = true,
 	drawtype = "nodebox",
+	leveled = 7,
 	node_box = {
-		type = "fixed",
-		fixed = {
-			{-0.5, -0.5, -0.5, 0.5, -0.25, 0.5},
-		},
+		type = "leveled",
+		fixed = {-0.5, -0.5, -0.5, 0.5, -0.5, 0.5},
 	},
 	collision_box = {
-		type = "fixed",
-		fixed = {
-			{-0.5, -0.5, -0.5, 0.5, -7 / 16, 0.5},
-		},
+		type = "leveled",
+		fixed = {-0.5, -0.5, -0.5, 0.5, -0.5, 0.5},
+	},
+	selection_box = {
+		type = "leveled",
+		fixed = {-0.5, -0.5, -0.5, 0.5, -0.5, 0.5},
 	},
 	groups = {crumbly = 3, falling_node = 1, snowy = 1},
 	sounds = default.node_sound_snow_defaults(),
 
 	on_construct = function(pos)
-		pos.y = pos.y - 1
-		if minetest.get_node(pos).name == "default:dirt_with_grass" then
-			minetest.set_node(pos, {name = "default:dirt_with_snow"})
+		pos.y = pos.y-1
+		local node = minetest.get_node(pos)
+		if node.name == "default:dirt_with_grass"
+		or node.name == "default:dirt" then
+			node.name = "default:dirt_with_snow"
+			minetest.set_node(pos, node)
+		end
+	end,
+
+	--Handle node drops due to node level.
+	on_dig = function(pos, node, digger)
+		local level = minetest.get_node_level(pos)
+		minetest.node_dig(pos, node, digger)
+		if minetest.get_node(pos).name ~= node.name then
+			local inv = digger:get_inventory()
+			if not inv then
+				return
+			end
+			local left = inv:add_item("main", "default:snow "..tostring(level/7-1))
+			if not left:is_empty() then
+				minetest.add_item({
+					x = pos.x + math.random()/2-0.25,
+					y = pos.y + math.random()/2-0.25,
+					z = pos.z + math.random()/2-0.25,
+				}, left)
+			end
+		end
+	end,
+
+	--Manage snow levels.
+	on_place = function(itemstack, player, pt)
+		local oldnode_under = minetest.get_node_or_nil(pt.under)
+		if not oldnode_under then
+			return itemstack, false
+		end
+
+		local olddef_under = minetest.registered_nodes[oldnode_under.name]
+		if not olddef_under then
+			return itemstack, false
+		end
+
+		-- If node under is buildable_to, place into it instead (eg. snow)
+		local pos, node
+		if olddef_under.buildable_to then
+			pos = pt.under
+			node = oldnode_under
+		else
+			pos = pt.above
+			node = minetest.get_node(pos)
+			local def = minetest.registered_nodes[node.name]
+			if not def
+			or not def.buildable_to then
+				return itemstack, false
+			end
+		end
+
+		-- nil player can place (for snowballs)
+		if player
+		and minetest.is_protected(pos, player:get_player_name()) then
+			return itemstack, false
 		end
+
+		if node.name ~= "default:snow" then
+			if minetest.get_node{x=pos.x, y=pos.y-1, z=pos.z}.name ==
+					"default:snow" then
+				-- grow the snow below (fixes levelled problem)
+				pos.y = pos.y - 1
+			else
+				-- place a snow
+				return minetest.item_place_node(itemstack, player, pt)
+			end
+		end
+
+		-- grow the snow
+		local level = minetest.get_node_level(pos)
+		level = level + 7
+		if level < 64 then
+			minetest.set_node_level(pos, level)
+		else
+			-- place a snowblock and snow onto it if possible
```

Comment:
I suggest to choose a snow level that is a multiple of one pixel, this removes the necessity to place another snow layer on top of the block.

Reply:
How would I do that?

Reply:
https://github.com/minetest/minetest/blob/95a37efc31586babadc2f42e310e13e8319158a3/doc/lua_api.txt#L967
>The nodebox height is (`param2` / 64) nodes.

One pixel is 1/16 node. Ergo the level step has to be 4 per pixel.

[Also see my ash mod.](https://github.com/DS-Minetest/ash/blob/58eba2e23444778d2e233d3595f36bb11a83a2bf/init.lua#L57)

Reply:
The leveled snow nodebox should only ever be multiples of 1/8th node.

Reply:
> I suggest to choose a snow level that is a multiple of one pixel, this removes the necessity to place another snow layer on top of the block.

I don't understand why.
If level becomes 63, full cube, change to snowblock.
Then the next placement of snow is just normal placement of snow on a snowblock.

Reply:
(If level becomes 63 it's only a 63/64 node, 64 would be correct here.)

If you look at the code, you can see, that the level of one snow layer is added and the remaining is added as another node. The code comment here `-- place a snowblock and snow onto it if possible` wants to tell us this.
This isn't needed to do if you choose a step level of 1/nth node. (Eg. 8 (= 2 pixels = 1/8th node), 4 (= 1 pixels = 1/16th node) or something else.)
It's wrong that it has to be a multiple of one pixel, sorry.

Reply:
Yes my mistake, i mean:
If level becomes 64, full cube, change to snowblock.